Integrate Split-Explicit Solver into Driver.jl

blallen opened the pull request:
# Description

See title. Draft for now. Main complication is dealing with a second balance, grid, and DGModel.

<!--- Please fill out the following section --->

I have

- [ ] Written and run all necessary tests with CLIMA by including `tests/runtests.jl`
- [ ] Followed all necessary [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) and run `julia .dev/climaformat.jl .`
- [ ] Updated the documentation to reflect changes from this PR.

<!--- Please leave the following section --->

# For review by CLIMA developers

- [ ] There are no open pull requests for this already
- [ ] CLIMA developers with relevant expertise have been assigned to review this submission
- [ ] The code conforms to the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) and has consistent naming conventions. `julia .dev/format.jl` has been run in a separate commit.
- [ ] This code does what it is technically intended to do (all numerics make sense physically and/or computationally)

kpamnany commented on src/Common/ConfigTypes/ConfigTypes.jl:
```diff
@@ -9,12 +9,16 @@ export ClimateMachineConfigType,
+struct OceanSplitExplicitConfigType <: ClimateMachineConfigType end
 struct SingleStackConfigType <: ClimateMachineConfigType end
 
+## WHY ISN'T THIS IN THE DRIVER FOLDER!?
```
Yeah, this and `Utilities/Checkpoint` should really be in `Driver/`.

blallen replied:
Oh good, I was afraid there was going to be some arcane reason why this was better

kpamnany commented on src/Driver/driver_configs.jl:
```diff
@@ -339,6 +343,78 @@ function OceanBoxGCMConfiguration(
+    FT = Float64,
+    array_type = ClimateMachine.array_type(),
+    solver_type = ExplicitSolverType(
+        # solver_method = SplitExplicitLSRK2nSolver,
```
Assume you'll remove any commented lines of code before merging.

blallen replied:
Yeah, this was just to get the driver working. I probably need to add an entirely new solver type here. That's the second half of this PR.

blallen commented:
@kpamnany I think I've finished the Driver.jl specific portions of this PR. It gets through the initialization and throws a method error on a `boundary_state!` call. So more work to be done, but I think the basic structure is there.

@jkozdon I think we are way abusing the `DGModel.modeldata` field with this setup, but will it actually cause anything bad?

jkozdon commented:
> @jkozdon I think we are way abusing the `DGModel.modeldata` field with this setup, but will it actually cause anything bad?

The `DGModel` was meant to be abused...

I don't think so though, since the kernels never touch the `modeldata`. Actually, I think the ocean is the only code that uses it:
```
❯ rg modeldata
experiments/OceanBoxGCM/simple_box.jl
71:    modeldata = (vert_filter = vert_filter, exp_filter = exp_filter)
81:        modeldata = modeldata,

src/Driver/solver_configs.jl
56:        modeldata=nothing,
73:# - `modeldata=nothing`: passed through to `DGModel`.
91:    modeldata = nothing,
130:            modeldata = modeldata,
149:            modeldata = modeldata,

test/Ocean/SplitExplicit/split_explicit.jl
69:    modeldata = (
112:            modeldata = modeldata,
137:            modeldata = modeldata,

test/Ocean/HydrostaticBoussinesq/test_3D_spindown.jl
88:    modeldata = (vert_filter = vert_filter, exp_filter = exp_filter)
97:        modeldata = modeldata,

test/Ocean/SplitExplicit/simple_box_ivd.jl
261:            # (dg.modeldata.tendency_dg.state_auxiliary,"tend Int aux",),
262:            # (dg.modeldata.conti3d_Q,"conti3d_Q",),

test/Ocean/SplitExplicit/simple_box_2dt.jl
257:            # (dg.modeldata.tendency_dg.state_auxiliary,"tend Int aux",),
258:            # (dg.modeldata.conti3d_Q,"conti3d_Q",),

src/Ocean/SplitExplicit/HydrostaticBoussinesqCoupling.jl
50:    model_int = dg.modeldata.integral_model

src/Ocean/SplitExplicit01/SplitExplicitLSRK2nMethod.jl
219:        ivdc_dg = slow.rhs!.modeldata.ivdc_dg
221:        ivdc_Q = slow.rhs!.modeldata.ivdc_Q
222:        ivdc_solver = slow.rhs!.modeldata.ivdc_bgm_solver
230:        ivdc_RHS = slow.rhs!.modeldata.ivdc_RHS

src/Ocean/SplitExplicit01/OceanModel.jl
137:    modeldata = (
157:        modeldata = modeldata,
476:    MD = dg.modeldata
494:    conti3d_dg = dg.modeldata.conti3d_dg
495:    # ct3d_Q = dg.modeldata.conti3d_Q
499:    ct3d_dQ = dg.modeldata.conti3d_Q
539:    flowintegral_dg = dg.modeldata.flowintegral_dg

src/Ocean/SplitExplicit01/Communication.jl
79:    tendency_dg = dgSlow.modeldata.tendency_dg
180:    flowintegral_dg = dgSlow.modeldata.flowintegral_dg

src/Ocean/SplitExplicit/Communication.jl
27:    model_int = model_bc.modeldata.integral_model
91:    model_int = model_bc.modeldata.integral_model

src/Ocean/HydrostaticBoussinesq/HydrostaticBoussinesqModel.jl
662:    MD = dg.modeldata

src/Numerics/DGMethods/DGModel.jl
15:    modeldata::MD
38:    modeldata = nothing,
53:        modeldata,

src/Numerics/DGMethods/remainder.jl
73:        modeldata,
102:    modeldata = maindg.modeldata,
155:        modeldata,
```

kpamnany commented on src/Driver/ConfigTypes/ConfigTypes.jl:
```diff
@@ -9,12 +9,16 @@ export ClimateMachineConfigType,
+struct OceanSplitExplicitConfigType <: ClimateMachineConfigType end
 struct SingleStackConfigType <: ClimateMachineConfigType end
 
+## WHY ISN'T THIS IN THE DRIVER FOLDER!?
+
```
```suggestion
```

blallen replied:
yep I'll remove it :)

kpamnany commented on src/Driver/solver_configs.jl:
```diff
@@ -118,17 +118,23 @@ function SolverConfiguration(
-            diffusion_direction = diffdir,
-            modeldata = modeldata,
-        )
+        if hasproperty(driver_config.config_info, :dg)
```
This makes me uncomfortable, but it'll get fixed when #1551 is implemented so I guess it's fine for now.

blallen replied:
This is me completely hacking in #1551 like functionality. After seeing how SolverTypes work, I might be able to move most of this to the SplitExplicitSolverType for now? Let me know what you think.

kpamnany replied:
I think, as part of addressing #1551, we can introduce `SolverConfiguration` constructors specialized on `driver_config.config_type`, as those constructors will basically become wrappers for `solversetup`.

blallen commented:
> > @jkozdon I think we are way abusing the `DGModel.modeldata` field with this setup, but will it actually cause anything bad?
> 
> The `DGModel` was meant to be abused...
> 
> I don't think so though, since the kernels never touch the `modeldata`. Actually, I think the ocean is the only code that uses it:

Got it. Will doing things like have the 2D balance law have the 3D balance law as a member and then storing the 2D DGModel in the 3D DGModel's model data cause unnecessary memory duplication? Because that's what we're currently doing and we can at least clean up the balance law end.

blallen commented:
bors try

blallen commented:
bors try

blallen commented:
@jkozdon @kpamnany any clue why `gpu_test_simple_box` and `gpu_simple_box_ivd` are initializing with what looks like NaNs? when I run on a local GPU everything is fine

https://buildkite.com/clima/climatemachine-ci/builds/370#acf0ddf1-5388-4396-bfca-c71e176dc521
https://buildkite.com/clima/climatemachine-ci/builds/367#86393589-0c92-4164-b3a7-53975d3f367b

blallen commented:
bors try

jkozdon commented:
> @jkozdon @kpamnany any clue why `gpu_test_simple_box` and `gpu_simple_box_ivd` are initializing with what looks like NaNs? when I run on a local GPU everything is fine
> 
> https://buildkite.com/clima/climatemachine-ci/builds/370#acf0ddf1-5388-4396-bfca-c71e176dc521
> https://buildkite.com/clima/climatemachine-ci/builds/367#86393589-0c92-4164-b3a7-53975d3f367b

I have no idea. Sorry.

I would try initializing the MPI state array for `NaN`s locally as you might have just gotten lucky locally.

Alternatively, you could be missing a `wait` and you are getting hit with out of order execution on the GPU. (When you say locally, you mean on your laptop or the MIT V100s?)

blallen commented:
> > @jkozdon @kpamnany any clue why `gpu_test_simple_box` and `gpu_simple_box_ivd` are initializing with what looks like NaNs? when I run on a local GPU everything is fine
> > https://buildkite.com/clima/climatemachine-ci/builds/370#acf0ddf1-5388-4396-bfca-c71e176dc521
> > https://buildkite.com/clima/climatemachine-ci/builds/367#86393589-0c92-4164-b3a7-53975d3f367b
> 
> I have no idea. Sorry.
> 
> I would try initializing the MPI state array for `NaN`s locally as you might have just gotten lucky locally.
> 
> Alternatively, you could be missing a `wait` and you are getting hit with out of order execution on the GPU. (When you say locally, you mean on your laptop or the MIT V100s?)

MIT V100s / Titan Vs, worked on both of them. I have a feeling it's something about the initialization, but it's initializing fine for me.

@kpamnany is the environment variable for `init_with_NaNs` set during the tests?

kpamnany commented:
[Nope](https://github.com/CliMA/ClimateMachine.jl/blob/master/.buildkite/pipeline.yml#L3).

blallen commented:
bors try

blallen commented:
bors try

blallen commented:
bors try-

blallen commented:
bors try

blallen commented:
@jkozdon seems to be an MPI effect. The test was previously using one task, I reverted to that and now it works

https://buildkite.com/clima/climatemachine-ci/builds/374#8d04bbb0-0058-4732-abdb-142adfd70ae7

the `test_simple_box.jl` test is hanging right now, but I can figure out why that is happening locally

blallen commented:
> @jkozdon seems to be an MPI effect. The test was previously using one task, I reverted to that and now it works
> 
> https://buildkite.com/clima/climatemachine-ci/builds/374#8d04bbb0-0058-4732-abdb-142adfd70ae7
> 
> the `test_simple_box.jl` test is hanging right now, but I can figure out why that is happening locally

turns out I modified the ntasks options for a different test by accident. should be good now

jkozdon commented:
> turns out I modified the ntasks options for a different test by accident. should be good now

Great. Though I am confused, shouldn't your code work in parallel? Why would it produce `NaN`s when run in parallel?

blallen commented:
> > turns out I modified the ntasks options for a different test by accident. should be good now
> 
> Great. Though I am confused, shouldn't your code work in parallel? Why would it produce `NaN`s when run in parallel?

It doesn't necessarily produce NaNs but the implicit iterative solve fails: https://buildkite.com/clima/climatemachine-ci/builds/374#105439a0-8550-4841-a4b4-451745f71369/373-407

Relevant code is: https://github.com/CliMA/ClimateMachine.jl/blob/master/src/Ocean/SplitExplicit01/SplitExplicitLSRK2nMethod.jl#L215-L260

blallen commented:
bors r+

jm-c commented on src/Driver/SolverTypes/SplitExplicitSolverType.jl:
```diff
@@ -0,0 +1,79 @@
+    fast_solver =
+        ode_solver.fast_method(dg_2D, Q_2D, dt = ode_solver.dt_fast, t0 = t0)
+    slow_solver =
+        ode_solver.fast_method(dg_3D, Q_3D, dt = ode_solver.dt_slow, t0 = t0)
```
Are we sure of this ? I would have thought that:
slow_solver = ode_solver.slow_method( ...) instead

blallen replied:
ah you are correct! this didn't get caught because we are currently using the same solver for both. I'll open up a bug fix PR for this change

blallen replied:
see #1577
